proposed/http-middleware/middleware-meta.md

```diff
@@ -136,59 +140,146 @@ only the request being passed to the middleware.
+response body has not been written to, which can lead to incomplete output or
+error responses being sent with cache headers attached. It is also possible
+to end up with [corrupted body content][rob-allen-filtering] when writing over
+existing body content if the new content is shorter than the original. The most
+effective way to resolve these issues is to always provide a fresh stream when
+modifying the body of a message.
+
+[rob-allen-filtering]: https://akrabat.com/filtering-the-psr-7-body-in-middleware/
+
+Some have argued that passing the response helps ensure dependency inversion.
+While it is true that it helps avoid depending on a specific implementation of
+HTTP messages, the problem can also be resolved by injecting factories into the
+middleware to create HTTP message objects, or by injecting empty message instances.
+With the creation of HTTP Factories in [PSR-17][psr17], a standard approach to
+handling dependency inversion is possible.
+
+[psr17]: https://github.com/php-fig/fig-standards/blob/master/proposed/http-factory/http-factory-meta.md
+
+A more subjective, but also important, concern is that existing double-pass
+middleware typically uses the `callable` type hint to refer to middleware.
+This makes strict typing impossible, as there is no assurance that the `callable`
+being passed implements a middleware signature, which reduces runtime safety.
+
+**Due to these significant issues the lambda approach has been choosen for this proposal.**
+
+5. Design Decisions
 -------------------
 
-The `$next` argument is a callable in most existing middleware systems. However using
-an interface allows to improve type safety and IDE support.
+### 5.1 Middleware Design
+
+The `MiddlewareInterface` defines a single method that accepts a server request
+and a delegate and must return a response. The middleware may:
+
+- Evolve the request before passing it to the delegate to execute the next
+  available middleware.
+- Evolve the response received from the delegate before returning it.
+- Create and return a response without passing it to the delegate, thereby
+  preventing any further middleware from executing.
+
+#### Why doesn't middleware use `__invoke`?
+
+Doing so would conflict with existing middleware that implements the double-pass
+approach and may want to implement the middleware interface.
+
+In addition, classes that define `__invoke` can be type hinted as `callable`,
+which results in less strict typing. This is generally undesirable, especially
+when the `__invoke` method uses strict typing.
+
+#### Why is a server request required?
+
+To make it clear that the middleware can only be used in a synchronous, server
+side context.
+
+While not all middleware will need to use the additional methods defined by the
+server request interface, external requests are typically handled asynchronously
+and would need to return a [promise][promises] of a response. (This is primarily
```

Comment:
You're right, "typically" should be used here, since that is what all (?) current async PHP frameworks do.